benches/benches/bevy_ecs/schedule.rs

```diff
@@ -0,0 +1,62 @@
+use bevy_app::App;
+use bevy_ecs::prelude::*;
+use criterion::{criterion_group, criterion_main, Criterion};
+
+criterion_group!(benches, build_schedule);
+criterion_main!(benches);
+
+fn build_schedule(criterion: &mut Criterion) {
+    // empty system
+    fn sys() {}
+
+    // Use multiple different kinds of label to ensure that dynamic dispatch
+    // doesn't somehow get optimized away.
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, SystemLabel)]
+    struct NumLabel(usize);
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, SystemLabel)]
+    struct DummyLabel;
+
+    let mut group = criterion.benchmark_group("build_schedule");
+    group.warm_up_time(std::time::Duration::from_millis(500));
+    group.measurement_time(std::time::Duration::from_secs(15));
+
+    // Method: generate a set of `graph_size` systems which have a One True Ordering.
+    // Add system to the stage with full constraints. Hopefully this should be maximimally
+    // difficult for bevy to figure out.
+    let labels: Vec<_> = (0..1000).map(NumLabel).collect();
+
+    // Benchmark graphs of different sizes.
+    for graph_size in [100, 500, 1000] {
+        // Basic benchmark without constraints.
+        group.bench_function(format!("{graph_size}_schedule_noconstraints"), |bencher| {
+            bencher.iter(|| {
+                let mut app = App::new();
+                for _ in 0..graph_size {
+                    app.add_system(sys);
+                }
+                app.run();
+            });
+        });
+
+        // Benchmark with constraints.
+        group.bench_function(format!("{graph_size}_schedule"), |bencher| {
+            bencher.iter(|| {
+                let mut app = App::new();
+                app.add_system(sys.label(DummyLabel));
+                for i in 0..graph_size {
```

Comment:
This should have a comment describing the strategy used to build the graph.

On first read it looks like we have a maximally dense graph, such that each system starts in a distinct order.

Personally I think this strategy is suboptimal; we should consider having a linear number of edges, so we can spot any emergent nonlinearities in the rest of the algorithm more easily.

In real use cases, I would expect the number of edges to grow somewhere between linearly and quadratically with the number of systems.